### PR TITLE
Add otlp log and trace exporter benchmarks

### PR DESCRIPTION
**Pull request by vishweshbankwar**
Fixes #
Design discussion issue #

## Changes

Please provide a brief description of the changes here.

## Merge requirement checklist

* [ ] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (nullable enabled, static analysis, etc.)
* [ ] Unit tests added/updated
* [ ] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

File: test/Benchmarks/Exporter/OtlpLogExporterBenchmarks.cs [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+BenchmarkDotNet v0.13.6, Windows 11 (10.0.22621.2134/22H2/2022Update/SunValley2) (Hyper-V)
+AMD EPYC 7763, 1 CPU, 16 logical and 8 physical cores
+.NET SDK 7.0.400
+  [Host]     : .NET 7.0.10 (7.0.1023.36312), X64 RyuJIT AVX2
+  DefaultJob : .NET 7.0.10 (7.0.1023.36312), X64 RyuJIT AVX2
+
+
+|            Method |       Mean |     Error |    StdDev |   Gen0 |   Gen1 | Allocated |
+|------------------ |-----------:|----------:|----------:|-------:|-------:|----------:|
+| OtlpExporter_Http | 147.112 us | 2.9354 us | 2.4512 us | 0.7324 | 0.4883 |   12289 B |
```
Comment (cijothomas) [inline test/Benchmarks/Exporter/OtlpLogExporterBenchmarks.cs:44]: nit: please run the same for traces and update the file with that numbers. so we can get a quick of how logexporter compares to traceexporter in terms of perf/alloc.

Comment (vishweshbankwar) [reply test/Benchmarks/Exporter/OtlpLogExporterBenchmarks.cs:44]: added

File: test/Benchmarks/Exporter/OtlpTraceExporterBenchmarks.cs [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    private CircularBuffer<Activity> activityBatch;
+
+    private IDisposable server;
+    private string serverHost;
+    private int serverPort;
+
+    [GlobalSetup(Target = nameof(OtlpExporter_Grpc))]
+    public void GlobalSetupGrpc()
+    {
+        var mockClient = new Mock<OtlpCollector.TraceService.TraceServiceClient>();
```
Comment (utpilla) [inline test/Benchmarks/Exporter/OtlpTraceExporterBenchmarks.cs:62]: Let's not add new code using `Moq`.

File: test/Benchmarks/Exporter/OtlpLogExporterBenchmarks.cs [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+            .Returns(new OtlpCollector.ExportLogsServiceResponse());
+
+        var options = new OtlpExporterOptions();
+        this.exporter = new OtlpLogExporter(
+            options,
+            new SdkLimitOptions(),
+            new OtlpGrpcLogExportClient(options, mockClient.Object));
+
+        this.logRecord = LogRecordHelper.CreateTestLogRecord();
+        this.logRecordBatch = new CircularBuffer<LogRecord>(1);
```
Comment (utpilla) [inline test/Benchmarks/Exporter/OtlpLogExporterBenchmarks.cs:78]: Add the log record to `logRecordBatch`.

File: test/Benchmarks/Exporter/OtlpTraceExporterBenchmarks.cs [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        mockClient
+            .Setup(m => m.Export(
+                It.IsAny<OtlpCollector.ExportTraceServiceRequest>(),
+                It.IsAny<Metadata>(),
+                It.IsAny<DateTime?>(),
+                It.IsAny<CancellationToken>()))
+            .Returns(new OtlpCollector.ExportTraceServiceResponse());
```
Comment (utpilla) [inline test/Benchmarks/Exporter/OtlpTraceExporterBenchmarks.cs:69]: This seems incorrect. Does the response returned actually contain the activity data that we want?

Comment (utpilla) [inline test/Benchmarks/Exporter/OtlpTraceExporterBenchmarks.cs:69]: This seems incorrect. Is it bypassing the whole logic of processing the activity and simply returning a new instance of `ExportTraceServiceResponse`?

Comment (utpilla) [reply test/Benchmarks/Exporter/OtlpTraceExporterBenchmarks.cs:69]: Isn't this skipping the grpc Unary call altogether?

Comment (vishweshbankwar) [reply test/Benchmarks/Exporter/OtlpTraceExporterBenchmarks.cs:69]: Yes, that is correct, removing moq should fix this.

File: test/Benchmarks/Exporter/OtlpTraceExporterBenchmarks.cs [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+    public void GlobalSetupHttp()
+    {
+        this.server = TestHttpServer.RunServer(
+            (ctx) =>
+            {
+                using (Stream receiveStream = ctx.Request.InputStream)
+                {
+                    while (true)
+                    {
+                        if (receiveStream.Read(this.buffer, 0, this.buffer.Length) == 0)
```
Comment (utpilla) [inline test/Benchmarks/Exporter/OtlpTraceExporterBenchmarks.cs:92]: Why do we need this?

File: test/Benchmarks/Exporter/OtlpTraceExporterBenchmarks.cs [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+
+|                 Method |     Mean |   Error |  StdDev |   Gen0 |   Gen1 | Allocated |
+|----------------------- |---------:|--------:|--------:|-------:|-------:|----------:|
+| OtlpTraceExporter_Http | 139.4 us | 1.41 us | 1.32 us | 0.4883 | 0.2441 |    9.8 KB |
+| OtlpTraceExporter_Grpc | 263.0 us | 3.47 us | 3.24 us | 0.4883 |      - |   9.34 KB |
+*/
+
+public class OtlpTraceExporterBenchmarks
+{
+    private readonly byte[] buffer = new byte[1024 * 1024];
```
Comment (utpilla) [inline test/Benchmarks/Exporter/OtlpTraceExporterBenchmarks.cs:54]: You could remove this.

**Comment by reyang** [comment]
I have a general comment which doesn't have to block this PR: how do we consider the granularity of the Benchmark project? (take Stress Test for example, we don't shove all components' test logic into a single test project)

**Comment by vishweshbankwar** [comment]
> I have a general comment which doesn't have to block this PR: how do we consider the granularity of the Benchmark project? (take Stress Test for example, we don't shove all components' test logic into a single test project)

@reyang We do not have a set rule for deciding granularity. Not sure if we have a need for individual projects for benchmarks as its easier to use filtering during execution.

File: test/Benchmarks/Exporter/OtlpLogExporterBenchmarks.cs [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+        this.exporter.Export(new Batch<LogRecord>(this.logRecordBatch, 1));
+    }
+
+    [Benchmark]
+    public void OtlpLogExporter_Grpc()
+    {
+        this.exporter.Export(new Batch<LogRecord>(this.logRecordBatch, 1));
+    }
+
+    private class MockLogService : OtlpCollector.LogsService.LogsServiceBase
```
Comment (utpilla) [inline test/Benchmarks/Exporter/OtlpLogExporterBenchmarks.cs:148]: nit:
```suggestion
    private sealed class MockLogService : OtlpCollector.LogsService.LogsServiceBase
```

File: test/Benchmarks/Exporter/OtlpTraceExporterBenchmarks.cs [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+        this.exporter.Export(new Batch<Activity>(this.activityBatch, 1));
+    }
+
+    [Benchmark]
+    public void OtlpTraceExporter_Grpc()
+    {
+        this.exporter.Export(new Batch<Activity>(this.activityBatch, 1));
+    }
+
+    private class MockTraceService : OtlpCollector.TraceService.TraceServiceBase
```
Comment (utpilla) [inline test/Benchmarks/Exporter/OtlpTraceExporterBenchmarks.cs:148]: nit:
```suggestion
    private sealed class MockTraceService : OtlpCollector.TraceService.TraceServiceBase
```

File: test/Benchmarks/Exporter/OtlpTraceExporterBenchmarks.cs [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+        this.exporter.Dispose();
+        this.host.Dispose();
+    }
+
+    [GlobalCleanup(Target = nameof(OtlpTraceExporter_Http))]
+    public void GlobalCleanupHttp()
+    {
+        this.exporter.Shutdown();
+        this.exporter.Dispose();
+        this.server.Dispose();
```
Comment (utpilla) [inline test/Benchmarks/Exporter/OtlpTraceExporterBenchmarks.cs:133]: You could dispose the activity as well.

File: test/Benchmarks/Exporter/OtlpLogExporterBenchmarks.cs [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+                ctx.Response.OutputStream.Close();
+            },
+            out this.serverHost,
+            out this.serverPort);
+
+        var options = new OtlpExporterOptions
+        {
+            Endpoint = new Uri($"http://{this.serverHost}:{this.serverPort}"),
+        };
+        this.exporter = new OtlpLogExporter(
```
Comment (utpilla) [inline test/Benchmarks/Exporter/OtlpLogExporterBenchmarks.cs:110]: Why use the internal overload of `OtlpLogExporter` ctor here? Could we be consistent with the Grpc benchmark and just use:
```csharp
var options = new OtlpExporterOptions
{
    Endpoint = new Uri($"http://{this.serverHost}:{this.serverPort}"),
};
this.exporter = new OtlpLogExporter(options);
```

Comment (vishweshbankwar) [reply test/Benchmarks/Exporter/OtlpLogExporterBenchmarks.cs:110]: missed that. fixed it

**Review by utpilla — APPROVED** [review]

**Pull request merged**